// @navikt/ds-react/cjs/tooltip/Tooltip.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tooltip = void 0;
const react_1 = require("@floating-ui/react");
const clsx_1 = __importDefault(require("clsx"));
const react_2 = __importStar(require("react"));
const Modal_context_1 = require("../modal/Modal.context");
const portal_1 = require("../portal");
const typography_1 = require("../typography");
const hooks_1 = require("../util/hooks");
const useControllableState_1 = require("../util/hooks/useControllableState");
const useMergeRefs_1 = require("../util/hooks/useMergeRefs");
/**
 * A component that displays a tooltip when the user hovers over its child element.
 *
 * @see [ðŸ“ Documentation](https://aksel.nav.no/komponenter/core/tooltip)
 * @see ðŸ·ï¸ {@link TooltipProps}
 *
 * @example
 * ```jsx
 * <Tooltip content="Skriv ut dokument">
 *   <Button icon={<PrinterLargeIcon title="demo knapp" />} />
 * </Tooltip>
 * ```
 */
exports.Tooltip = (0, react_2.forwardRef)((_a, ref) => {
    var { children, className, arrow: _arrow = true, placement: _placement = "top", open, defaultOpen = false, onOpenChange, offset: _offset, content, delay = 150, id, keys, maxChar = 80 } = _a, rest = __rest(_a, ["children", "className", "arrow", "placement", "open", "defaultOpen", "onOpenChange", "offset", "content", "delay", "id", "keys", "maxChar"]);
    const [_open, _setOpen] = (0, useControllableState_1.useControllableState)({
        defaultValue: defaultOpen,
        value: open,
        onChange: onOpenChange,
    });
    const arrowRef = (0, react_2.useRef)(null);
    const modalContext = (0, Modal_context_1.useModalContext)(false);
    const rootElement = modalContext ? modalContext.ref.current : undefined;
    const { x, y, strategy, context, placement, middlewareData: { arrow: { x: arrowX, y: arrowY } = {}, hide: { referenceHidden } = {}, }, refs, } = (0, react_1.useFloating)({
        placement: _placement,
        open: _open,
        onOpenChange: (newState) => _setOpen(newState),
        middleware: [
            (0, react_1.offset)(_offset ? _offset : _arrow ? 10 : 2),
            (0, react_1.shift)(),
            (0, react_1.flip)({ padding: 5, fallbackPlacements: ["bottom", "top"] }),
            (0, react_1.arrow)({ element: arrowRef, padding: 5 }),
        ],
        whileElementsMounted: modalContext
            ? (reference, floating, update) => 
            // Reduces jumping in Chrome when used in a Modal and it's the first focusable element.
            // Can be removed when autofocus starts working on <dialog> in Chrome. See also Modal.tsx
            (0, react_1.autoUpdate)(reference, floating, update, { animationFrame: true })
            : react_1.autoUpdate,
        strategy: modalContext ? "fixed" : undefined,
    });
    const { getReferenceProps, getFloatingProps } = (0, react_1.useInteractions)([
        (0, react_1.useHover)(context, { handleClose: (0, react_1.safePolygon)(), restMs: delay }),
        (0, react_1.useFocus)(context),
        (0, react_1.useDismiss)(context),
    ]);
    const ariaId = (0, hooks_1.useId)(id);
    const mergedRef = (0, useMergeRefs_1.useMergeRefs)(ref, refs.setFloating);
    const childMergedRef = (0, useMergeRefs_1.useMergeRefs)(children.ref, refs.setReference);
    if (!children ||
        (children === null || children === void 0 ? void 0 : children.type) === react_2.default.Fragment ||
        children === react_2.default.Fragment) {
        console.error("<Tooltip> children needs to be a single ReactElement and not: <React.Fragment/> | <></>");
        return null;
    }
    if ((content === null || content === void 0 ? void 0 : content.length) > maxChar) {
        _open &&
            console.warn(`Because of strict accessibility concers we encourage all Tooltips to have less than 80 characters. Can be overwritten with the maxChar-prop\n\nLength:${content.length}\nTooltip-content: ${content}`);
    }
    return (react_2.default.createElement(react_2.default.Fragment, null,
        (0, react_2.cloneElement)(children, getReferenceProps(Object.assign(Object.assign({}, children.props), { ref: childMergedRef, "aria-describedby": _open
                ? (0, clsx_1.default)(ariaId, children === null || children === void 0 ? void 0 : children.props["aria-describedby"])
                : children === null || children === void 0 ? void 0 : children.props["aria-describedby"] }))),
        react_2.default.createElement(portal_1.Portal, { rootElement: rootElement, asChild: true }, _open && (react_2.default.createElement("div", Object.assign({}, getFloatingProps(Object.assign(Object.assign({}, rest), { ref: mergedRef, style: {
                position: strategy,
                top: y !== null && y !== void 0 ? y : 0,
                left: x !== null && x !== void 0 ? x : 0,
                visibility: referenceHidden ? "hidden" : "visible",
            }, role: "tooltip", id: ariaId, className: (0, clsx_1.default)("navds-tooltip", "navds-detail navds-detail--small", className) })), { "data-side": placement }),
            content,
            keys && (react_2.default.createElement("span", { className: "navds-tooltip__keys" }, keys.map((key) => (react_2.default.createElement(typography_1.Detail, { as: "kbd", key: key, className: "navds-tooltip__key" }, key))))),
            _arrow && (react_2.default.createElement("div", { ref: (node) => {
                    arrowRef.current = node;
                }, className: "navds-tooltip__arrow", style: {
                    left: arrowX != null ? `${arrowX}px` : "",
                    top: arrowY != null ? `${arrowY}px` : "",
                    right: "",
                    bottom: "",
                    [{
                        top: "bottom",
                        right: "left",
                        bottom: "top",
                        left: "right",
                    }[placement]]: "-3.5px",
                } })))))));
});
exports.default = exports.Tooltip;
