// @navikt/ds-react/cjs/index.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Heading = exports.ErrorMessage = exports.Detail = exports.BodyShort = exports.BodyLong = exports.Tooltip = exports.ToggleGroup = exports.Timeline = exports.Tag = exports.Tabs = exports.Table = exports.Stepper = exports.Skeleton = exports.ReadMore = exports.Provider = exports.ProgressBar = exports.Portal = exports.Popover = exports.Pagination = exports.Modal = exports.Loader = exports.List = exports.Link = exports.VStack = exports.Stack = exports.Spacer = exports.HStack = exports.Show = exports.Hide = exports.Page = exports.HGrid = exports.Box = exports.Bleed = exports.InternalHeader = exports.HelpText = exports.GuidePanelDefaultIllustration = exports.GuidePanel = exports.ExpansionCard = exports.Dropdown = exports.useMonthpicker = exports.MonthPicker = exports.useRangeDatepicker = exports.useDatepicker = exports.DatePicker = exports.CopyButton = exports.Chips = exports.Chat = exports.Button = exports.Alert = exports.Accordion = void 0;
exports.Panel = exports.LinkPanel = exports.TextField = exports.Textarea = exports.Switch = exports.Select = exports.Search = exports.RadioGroup = exports.Radio = exports.FormProgress = exports.FormSummary = exports.UNSAFE_FileUpload = exports.Fieldset = exports.ErrorSummary = exports.ConfirmationPanel = exports.UNSAFE_Combobox = exports.CheckboxGroup = exports.Checkbox = exports.useId = exports.useEventListener = exports.useClientLayoutEffect = exports.omit = exports.debounce = exports.Label = exports.Ingress = void 0;
var accordion_1 = require("./accordion");
Object.defineProperty(exports, "Accordion", { enumerable: true, get: function () { return accordion_1.Accordion; } });
var alert_1 = require("./alert");
Object.defineProperty(exports, "Alert", { enumerable: true, get: function () { return alert_1.Alert; } });
var button_1 = require("./button");
Object.defineProperty(exports, "Button", { enumerable: true, get: function () { return button_1.Button; } });
var chat_1 = require("./chat");
Object.defineProperty(exports, "Chat", { enumerable: true, get: function () { return chat_1.Chat; } });
var chips_1 = require("./chips");
Object.defineProperty(exports, "Chips", { enumerable: true, get: function () { return chips_1.Chips; } });
var copybutton_1 = require("./copybutton");
Object.defineProperty(exports, "CopyButton", { enumerable: true, get: function () { return copybutton_1.CopyButton; } });
var datepicker_1 = require("./date/datepicker");
Object.defineProperty(exports, "DatePicker", { enumerable: true, get: function () { return datepicker_1.DatePicker; } });
Object.defineProperty(exports, "useDatepicker", { enumerable: true, get: function () { return datepicker_1.useDatepicker; } });
Object.defineProperty(exports, "useRangeDatepicker", { enumerable: true, get: function () { return datepicker_1.useRangeDatepicker; } });
var monthpicker_1 = require("./date/monthpicker");
Object.defineProperty(exports, "MonthPicker", { enumerable: true, get: function () { return monthpicker_1.MonthPicker; } });
Object.defineProperty(exports, "useMonthpicker", { enumerable: true, get: function () { return monthpicker_1.useMonthpicker; } });
var dropdown_1 = require("./dropdown");
Object.defineProperty(exports, "Dropdown", { enumerable: true, get: function () { return dropdown_1.Dropdown; } });
var expansion_card_1 = require("./expansion-card");
Object.defineProperty(exports, "ExpansionCard", { enumerable: true, get: function () { return expansion_card_1.ExpansionCard; } });
var guide_panel_1 = require("./guide-panel");
Object.defineProperty(exports, "GuidePanel", { enumerable: true, get: function () { return guide_panel_1.GuidePanel; } });
Object.defineProperty(exports, "GuidePanelDefaultIllustration", { enumerable: true, get: function () { return guide_panel_1.GuidePanelDefaultIllustration; } });
var help_text_1 = require("./help-text");
Object.defineProperty(exports, "HelpText", { enumerable: true, get: function () { return help_text_1.HelpText; } });
var internal_header_1 = require("./internal-header");
Object.defineProperty(exports, "InternalHeader", { enumerable: true, get: function () { return internal_header_1.InternalHeader; } });
var bleed_1 = require("./layout/bleed");
Object.defineProperty(exports, "Bleed", { enumerable: true, get: function () { return bleed_1.Bleed; } });
var box_1 = require("./layout/box");
Object.defineProperty(exports, "Box", { enumerable: true, get: function () { return box_1.Box; } });
var grid_1 = require("./layout/grid");
Object.defineProperty(exports, "HGrid", { enumerable: true, get: function () { return grid_1.HGrid; } });
var page_1 = require("./layout/page");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return page_1.Page; } });
var responsive_1 = require("./layout/responsive");
Object.defineProperty(exports, "Hide", { enumerable: true, get: function () { return responsive_1.Hide; } });
Object.defineProperty(exports, "Show", { enumerable: true, get: function () { return responsive_1.Show; } });
var stack_1 = require("./layout/stack");
Object.defineProperty(exports, "HStack", { enumerable: true, get: function () { return stack_1.HStack; } });
Object.defineProperty(exports, "Spacer", { enumerable: true, get: function () { return stack_1.Spacer; } });
Object.defineProperty(exports, "Stack", { enumerable: true, get: function () { return stack_1.Stack; } });
Object.defineProperty(exports, "VStack", { enumerable: true, get: function () { return stack_1.VStack; } });
var link_1 = require("./link");
Object.defineProperty(exports, "Link", { enumerable: true, get: function () { return link_1.Link; } });
var list_1 = require("./list");
Object.defineProperty(exports, "List", { enumerable: true, get: function () { return list_1.List; } });
var loader_1 = require("./loader");
Object.defineProperty(exports, "Loader", { enumerable: true, get: function () { return loader_1.Loader; } });
var modal_1 = require("./modal");
Object.defineProperty(exports, "Modal", { enumerable: true, get: function () { return modal_1.Modal; } });
var pagination_1 = require("./pagination");
Object.defineProperty(exports, "Pagination", { enumerable: true, get: function () { return pagination_1.Pagination; } });
var popover_1 = require("./popover");
Object.defineProperty(exports, "Popover", { enumerable: true, get: function () { return popover_1.Popover; } });
var portal_1 = require("./portal");
Object.defineProperty(exports, "Portal", { enumerable: true, get: function () { return portal_1.Portal; } });
var progress_bar_1 = require("./progress-bar");
Object.defineProperty(exports, "ProgressBar", { enumerable: true, get: function () { return progress_bar_1.ProgressBar; } });
var provider_1 = require("./provider");
Object.defineProperty(exports, "Provider", { enumerable: true, get: function () { return provider_1.Provider; } });
var read_more_1 = require("./read-more");
Object.defineProperty(exports, "ReadMore", { enumerable: true, get: function () { return read_more_1.ReadMore; } });
var skeleton_1 = require("./skeleton");
Object.defineProperty(exports, "Skeleton", { enumerable: true, get: function () { return skeleton_1.Skeleton; } });
var stepper_1 = require("./stepper");
Object.defineProperty(exports, "Stepper", { enumerable: true, get: function () { return stepper_1.Stepper; } });
var table_1 = require("./table");
Object.defineProperty(exports, "Table", { enumerable: true, get: function () { return table_1.Table; } });
var tabs_1 = require("./tabs");
Object.defineProperty(exports, "Tabs", { enumerable: true, get: function () { return tabs_1.Tabs; } });
var tag_1 = require("./tag");
Object.defineProperty(exports, "Tag", { enumerable: true, get: function () { return tag_1.Tag; } });
var timeline_1 = require("./timeline");
Object.defineProperty(exports, "Timeline", { enumerable: true, get: function () { return timeline_1.Timeline; } });
var toggle_group_1 = require("./toggle-group");
Object.defineProperty(exports, "ToggleGroup", { enumerable: true, get: function () { return toggle_group_1.ToggleGroup; } });
var tooltip_1 = require("./tooltip");
Object.defineProperty(exports, "Tooltip", { enumerable: true, get: function () { return tooltip_1.Tooltip; } });
var typography_1 = require("./typography");
Object.defineProperty(exports, "BodyLong", { enumerable: true, get: function () { return typography_1.BodyLong; } });
Object.defineProperty(exports, "BodyShort", { enumerable: true, get: function () { return typography_1.BodyShort; } });
Object.defineProperty(exports, "Detail", { enumerable: true, get: function () { return typography_1.Detail; } });
Object.defineProperty(exports, "ErrorMessage", { enumerable: true, get: function () { return typography_1.ErrorMessage; } });
Object.defineProperty(exports, "Heading", { enumerable: true, get: function () { return typography_1.Heading; } });
Object.defineProperty(exports, "Ingress", { enumerable: true, get: function () { return typography_1.Ingress; } });
Object.defineProperty(exports, "Label", { enumerable: true, get: function () { return typography_1.Label; } });
var util_1 = require("./util");
Object.defineProperty(exports, "debounce", { enumerable: true, get: function () { return util_1.debounce; } });
Object.defineProperty(exports, "omit", { enumerable: true, get: function () { return util_1.omit; } });
Object.defineProperty(exports, "useClientLayoutEffect", { enumerable: true, get: function () { return util_1.useClientLayoutEffect; } });
Object.defineProperty(exports, "useEventListener", { enumerable: true, get: function () { return util_1.useEventListener; } });
Object.defineProperty(exports, "useId", { enumerable: true, get: function () { return util_1.useId; } });
var checkbox_1 = require("./form/checkbox");
Object.defineProperty(exports, "Checkbox", { enumerable: true, get: function () { return checkbox_1.Checkbox; } });
Object.defineProperty(exports, "CheckboxGroup", { enumerable: true, get: function () { return checkbox_1.CheckboxGroup; } });
var combobox_1 = require("./form/combobox");
Object.defineProperty(exports, "UNSAFE_Combobox", { enumerable: true, get: function () { return combobox_1.UNSAFE_Combobox; } });
var confirmation_panel_1 = require("./form/confirmation-panel");
Object.defineProperty(exports, "ConfirmationPanel", { enumerable: true, get: function () { return confirmation_panel_1.ConfirmationPanel; } });
var error_summary_1 = require("./form/error-summary");
Object.defineProperty(exports, "ErrorSummary", { enumerable: true, get: function () { return error_summary_1.ErrorSummary; } });
var fieldset_1 = require("./form/fieldset");
Object.defineProperty(exports, "Fieldset", { enumerable: true, get: function () { return fieldset_1.Fieldset; } });
var file_upload_1 = require("./form/file-upload");
Object.defineProperty(exports, "UNSAFE_FileUpload", { enumerable: true, get: function () { return file_upload_1.UNSAFE_FileUpload; } });
var form_summary_1 = require("./form/form-summary");
Object.defineProperty(exports, "FormSummary", { enumerable: true, get: function () { return form_summary_1.FormSummary; } });
var form_progress_1 = require("./form/form-progress");
Object.defineProperty(exports, "FormProgress", { enumerable: true, get: function () { return form_progress_1.FormProgress; } });
var radio_1 = require("./form/radio");
Object.defineProperty(exports, "Radio", { enumerable: true, get: function () { return radio_1.Radio; } });
Object.defineProperty(exports, "RadioGroup", { enumerable: true, get: function () { return radio_1.RadioGroup; } });
var search_1 = require("./form/search");
Object.defineProperty(exports, "Search", { enumerable: true, get: function () { return search_1.Search; } });
var select_1 = require("./form/select");
Object.defineProperty(exports, "Select", { enumerable: true, get: function () { return select_1.Select; } });
var switch_1 = require("./form/switch");
Object.defineProperty(exports, "Switch", { enumerable: true, get: function () { return switch_1.Switch; } });
var textarea_1 = require("./form/textarea");
Object.defineProperty(exports, "Textarea", { enumerable: true, get: function () { return textarea_1.Textarea; } });
var textfield_1 = require("./form/textfield");
Object.defineProperty(exports, "TextField", { enumerable: true, get: function () { return textfield_1.TextField; } });
/**
 * @deprecated
 */
var link_panel_1 = require("./link-panel");
Object.defineProperty(exports, "LinkPanel", { enumerable: true, get: function () { return link_panel_1.LinkPanel; } });
var panel_1 = require("./panel");
Object.defineProperty(exports, "Panel", { enumerable: true, get: function () { return panel_1.Panel; } });
