// @navikt/ds-react/cjs/toggle-group/ToggleGroup.context.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.useToggleGroupContext = exports.ToggleGroupProvider = exports.useToggleGroupDescendant = exports.useToggleGroupDescendants = exports.useToggleGroupDescendantsContext = exports.ToggleGroupDescendantsProvider = exports.ToggleGroupContext = void 0;
const react_1 = require("react");
const create_context_1 = require("../util/create-context");
const useDescendant_1 = require("../util/hooks/descendants/useDescendant");
exports.ToggleGroupContext = (0, react_1.createContext)(null);
_a = (0, useDescendant_1.createDescendantContext)(), exports.ToggleGroupDescendantsProvider = _a[0], exports.useToggleGroupDescendantsContext = _a[1], exports.useToggleGroupDescendants = _a[2], exports.useToggleGroupDescendant = _a[3];
/* State context */
_b = (0, create_context_1.createContext)({
    name: "ToggleGroupContext",
    hookName: "useToggleGroupContext",
    providerName: "ToggleGroupProvider",
    errorMessage: "<ToggleGroup.Item> needs to be wrapped within <ToggleGroup>",
}), exports.ToggleGroupProvider = _b[0], exports.useToggleGroupContext = _b[1];
