// @navikt/ds-react/cjs/form/form-summary/index.js
"use strict";
"use client";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormSummaryValue = exports.FormSummaryLabel = exports.FormSummaryHeading = exports.FormSummaryHeader = exports.FormSummaryEditLink = exports.FormSummaryAnswers = exports.FormSummaryAnswer = exports.FormSummary = void 0;
var FormSummary_1 = require("./FormSummary");
Object.defineProperty(exports, "FormSummary", { enumerable: true, get: function () { return __importDefault(FormSummary_1).default; } });
var FormSummaryAnswer_1 = require("./FormSummaryAnswer");
Object.defineProperty(exports, "FormSummaryAnswer", { enumerable: true, get: function () { return __importDefault(FormSummaryAnswer_1).default; } });
var FormSummaryAnswers_1 = require("./FormSummaryAnswers");
Object.defineProperty(exports, "FormSummaryAnswers", { enumerable: true, get: function () { return __importDefault(FormSummaryAnswers_1).default; } });
var FormSummaryEditLink_1 = require("./FormSummaryEditLink");
Object.defineProperty(exports, "FormSummaryEditLink", { enumerable: true, get: function () { return __importDefault(FormSummaryEditLink_1).default; } });
var FormSummaryHeader_1 = require("./FormSummaryHeader");
Object.defineProperty(exports, "FormSummaryHeader", { enumerable: true, get: function () { return __importDefault(FormSummaryHeader_1).default; } });
var FormSummaryHeading_1 = require("./FormSummaryHeading");
Object.defineProperty(exports, "FormSummaryHeading", { enumerable: true, get: function () { return __importDefault(FormSummaryHeading_1).default; } });
var FormSummaryLabel_1 = require("./FormSummaryLabel");
Object.defineProperty(exports, "FormSummaryLabel", { enumerable: true, get: function () { return __importDefault(FormSummaryLabel_1).default; } });
var FormSummaryValue_1 = require("./FormSummaryValue");
Object.defineProperty(exports, "FormSummaryValue", { enumerable: true, get: function () { return __importDefault(FormSummaryValue_1).default; } });
