// @navikt/ds-tokens/dist/tokens.js
/**
 * Do not edit directly
 * Generated on Wed, 19 Jun 2024 07:18:33 GMT
 */

export const ABreakpointXs = "0";
export const ABreakpointSm = "480px";
export const ABreakpointSmDown = "479px";
export const ABreakpointMd = "768px";
export const ABreakpointMdDown = "767px";
export const ABreakpointLg = "1024px";
export const ABreakpointLgDown = "1023px";
export const ABreakpointXl = "1280px";
export const ABreakpointXlDown = "1279px";
export const ABreakpoint2xl = "1440px";
export const ABreakpoint2xlDown = "1439px";
export const ATextWidthMax = "576px";
export const AFontFamily = "'Source Sans Pro', Arial, sans-serif";
export const AFontLineHeightHeading2xlarge = "3.25rem";
export const AFontLineHeightHeadingXlarge = "2.5rem";
export const AFontLineHeightHeadingLarge = "2.25rem";
export const AFontLineHeightHeadingMedium = "2rem";
export const AFontLineHeightHeadingSmall = "1.75rem";
export const AFontLineHeightHeadingXsmall = "1.5rem";
export const AFontLineHeightXlarge = "1.75rem";
export const AFontLineHeightLarge = "1.5rem";
export const AFontLineHeightMedium = "1.25rem";
export const AFontSizeHeading2xlarge = "2.5rem";
export const AFontSizeHeadingXlarge = "2rem";
export const AFontSizeHeadingLarge = "1.75rem";
export const AFontSizeHeadingMedium = "1.5rem";
export const AFontSizeHeadingSmall = "1.25rem";
export const AFontSizeHeadingXsmall = "1.125rem";
export const AFontSizeXlarge = "1.25rem";
export const AFontSizeLarge = "1.125rem";
export const AFontSizeMedium = "1rem";
export const AFontSizeSmall = "0.875rem";
export const AFontWeightBold = "600";
export const AFontWeightRegular = "400";
export const AZIndexPopover = 1000;
export const AZIndexFocus = 10;
export const AZIndexTooltip = 3000;
export const ABorderRadiusSmall = "2px";
export const ABorderRadiusMedium = "4px";
export const ABorderRadiusLarge = "8px";
export const ABorderRadiusXlarge = "12px";
export const ABorderRadiusFull = "9999px";
export const ABgDefault = "rgba(255, 255, 255, 1)";
export const ABgSubtle = "rgba(236, 238, 240, 1)";
export const ABorderSubtleHover = "rgba(5, 23, 51, 0.34)";
export const ABorderDefault = "rgba(2, 20, 49, 0.49)";
export const ABorderStrong = "rgba(1, 11, 24, 0.68)";
export const ABorderDivider = "rgba(7, 26, 54, 0.21)";
export const ABorderSubtle = "rgba(7, 26, 54, 0.21)";
export const ABorderActionSelected = "rgba(0, 69, 156, 1)";
export const ABorderActionHover = "rgba(0, 86, 180, 1)";
export const ABorderAction = "rgba(0, 103, 197, 1)";
export const ABorderSelected = "rgba(0, 103, 197, 1)";
export const ABorderSuccess = "rgba(6, 137, 58, 1)";
export const ABorderDanger = "rgba(195, 0, 0, 1)";
export const ABorderWarning = "rgba(199, 115, 0, 1)";
export const ABorderInfo = "rgba(54, 141, 168, 1)";
export const ABorderFocusOnInverted = "rgba(153, 195, 255, 1)";
export const ABorderFocus = "rgba(0, 52, 125, 1)";
export const ABorderOnInverted = "rgba(224, 227, 230, 1)";
export const ABorderOnInvertedSubtle = "rgba(82, 89, 98, 1)";
export const ABorderAlt1 = "rgba(130, 105, 162, 1)";
export const ABorderAlt2 = "rgba(127, 137, 0, 1)";
export const ABorderAlt3 = "rgba(0, 91, 130, 1)";
export const AIconDefault = "rgba(35, 38, 42, 1)";
export const AIconSubtle = "rgba(1, 11, 24, 0.68)";
export const AIconActionSelected = "rgba(0, 69, 156, 1)";
export const AIconActionHover = "rgba(0, 86, 180, 1)";
export const AIconActionOnActionSubtle = "rgba(0, 86, 180, 1)";
export const AIconAction = "rgba(0, 103, 197, 1)";
export const AIconSuccess = "rgba(6, 137, 58, 1)";
export const AIconDanger = "rgba(195, 0, 0, 1)";
export const AIconWarning = "rgba(199, 115, 0, 1)";
export const AIconInfo = "rgba(35, 107, 125, 1)";
export const AIconAlt1 = "rgba(99, 70, 137, 1)";
export const AIconAlt2 = "rgba(127, 137, 0, 1)";
export const AIconAlt3 = "rgba(0, 91, 130, 1)";
export const AIconOnNeutral = "rgba(255, 255, 255, 1)";
export const AIconOnInverted = "rgba(255, 255, 255, 1)";
export const AIconOnAction = "rgba(255, 255, 255, 1)";
export const AIconOnSuccess = "rgba(255, 255, 255, 1)";
export const AIconOnDanger = "rgba(255, 255, 255, 1)";
export const AIconOnWarning = "rgba(35, 38, 42, 1)";
export const AIconOnInfo = "rgba(35, 38, 42, 1)";
export const ASurfaceHover = "rgba(18, 43, 68, 0.08)";
export const ASurfaceActive = "rgba(17, 41, 64, 0.13)";
export const ASurfaceInvertedHover = "rgba(66, 71, 79, 1)";
export const ASurfaceInvertedActive = "rgba(82, 89, 98, 1)";
export const ASurfaceActionSubtleHover = "rgba(204, 225, 255, 1)";
export const ASurfaceActionHover = "rgba(0, 86, 180, 1)";
export const ASurfaceActionActive = "rgba(0, 69, 156, 1)";
export const ASurfaceActionSelectedHover = "rgba(0, 52, 125, 1)";
export const ASurfaceNeutralSubtleHover = "rgba(17, 41, 64, 0.13)";
export const ASurfaceNeutralActive = "rgba(35, 38, 42, 1)";
export const ASurfaceNeutralHover = "rgba(66, 71, 79, 1)";
export const ASurfaceSuccessSubtleHover = "rgba(153, 222, 173, 1)";
export const ASurfaceSuccessHover = "rgba(0, 124, 46, 1)";
export const ASurfaceDangerSubtleHover = "rgba(246, 130, 130, 1)";
export const ASurfaceDangerHover = "rgba(173, 0, 0, 1)";
export const ASurfaceDangerActive = "rgba(140, 0, 0, 1)";
export const ASurfaceWarningSubtleHover = "rgba(255, 215, 153, 1)";
export const ASurfaceInfoSubtleHover = "rgba(181, 241, 255, 1)";
export const ASurfaceDefault = "rgba(255, 255, 255, 1)";
export const ASurfaceSelected = "rgba(230, 240, 255, 1)";
export const ASurfaceSubtle = "rgba(242, 243, 245, 1)";
export const ASurfaceTransparent = "rgba(255, 255, 255, 0)";
export const ASurfaceBackdrop = "rgba(2, 20, 49, 0.49)";
export const ASurfaceInverted = "rgba(35, 38, 42, 1)";
export const ASurfaceActionSubtle = "rgba(230, 240, 255, 1)";
export const ASurfaceActionSelected = "rgba(0, 69, 156, 1)";
export const ASurfaceAction = "rgba(0, 103, 197, 1)";
export const ASurfaceNeutralSubtle = "rgba(18, 43, 68, 0.08)";
export const ASurfaceNeutralModerate = "rgba(17, 41, 64, 0.13)";
export const ASurfaceNeutralSelected = "rgba(35, 38, 42, 1)";
export const ASurfaceNeutral = "rgba(82, 89, 98, 1)";
export const ASurfaceSuccessSubtle = "rgba(204, 241, 214, 1)";
export const ASurfaceSuccessModerate = "rgba(153, 222, 173, 1)";
export const ASurfaceSuccess = "rgba(6, 137, 58, 1)";
export const ASurfaceDangerSubtle = "rgba(255, 194, 194, 1)";
export const ASurfaceDangerModerate = "rgba(246, 130, 130, 1)";
export const ASurfaceDanger = "rgba(195, 0, 0, 1)";
export const ASurfaceWarningSubtle = "rgba(255, 236, 204, 1)";
export const ASurfaceWarningModerate = "rgba(255, 215, 153, 1)";
export const ASurfaceWarning = "rgba(255, 145, 0, 1)";
export const ASurfaceInfoSubtle = "rgba(216, 249, 255, 1)";
export const ASurfaceInfoModerate = "rgba(181, 241, 255, 1)";
export const ASurfaceInfo = "rgba(102, 203, 236, 1)";
export const ASurfaceAlt1Subtle = "rgba(224, 216, 233, 1)";
export const ASurfaceAlt1Moderate = "rgba(192, 178, 210, 1)";
export const ASurfaceAlt1 = "rgba(130, 105, 162, 1)";
export const ASurfaceAlt2Subtle = "rgba(249, 252, 204, 1)";
export const ASurfaceAlt2Moderate = "rgba(236, 243, 153, 1)";
export const ASurfaceAlt2 = "rgba(193, 203, 51, 1)";
export const ASurfaceAlt3Subtle = "rgba(204, 226, 240, 1)";
export const ASurfaceAlt3Moderate = "rgba(153, 196, 221, 1)";
export const ASurfaceAlt3Strong = "rgba(0, 52, 83, 1)";
export const ASurfaceAlt3 = "rgba(0, 91, 130, 1)";
export const ATextDefault = "rgba(35, 38, 42, 1)";
export const ATextSubtle = "rgba(1, 11, 24, 0.68)";
export const ATextVisited = "rgba(99, 70, 137, 1)";
export const ATextDanger = "rgba(195, 0, 0, 1)";
export const ATextActionSelected = "rgba(0, 69, 156, 1)";
export const ATextActionHover = "rgba(0, 86, 180, 1)";
export const ATextActionOnActionSubtle = "rgba(0, 86, 180, 1)";
export const ATextAction = "rgba(0, 103, 197, 1)";
export const ATextOnInverted = "rgba(255, 255, 255, 1)";
export const ATextOnNeutral = "rgba(255, 255, 255, 1)";
export const ATextOnAction = "rgba(255, 255, 255, 1)";
export const ATextOnSuccess = "rgba(255, 255, 255, 1)";
export const ATextOnDanger = "rgba(255, 255, 255, 1)";
export const ATextOnWarning = "rgba(35, 38, 42, 1)";
export const ATextOnInfo = "rgba(35, 38, 42, 1)";
export const ATextOnAlt1 = "rgba(255, 255, 255, 1)";
export const ATextOnAlt2 = "rgba(35, 38, 42, 1)";
export const ATextOnAlt3 = "rgba(255, 255, 255, 1)";
export const ADataSurface1 = "rgba(51, 134, 224, 1)";
export const ADataSurface1Subtle = "rgba(204, 225, 255, 1)";
export const ADataSurface2 = "rgba(0, 91, 130, 1)";
export const ADataSurface2Subtle = "rgba(153, 196, 221, 1)";
export const ADataSurface3 = "rgba(199, 115, 0, 1)";
export const ADataSurface3Subtle = "rgba(255, 193, 102, 1)";
export const ADataSurface4 = "rgba(54, 141, 168, 1)";
export const ADataSurface4Subtle = "rgba(102, 203, 236, 1)";
export const ADataSurface5 = "rgba(51, 170, 95, 1)";
export const ADataSurface5Subtle = "rgba(153, 222, 173, 1)";
export const ADataSurface6 = "rgba(130, 105, 162, 1)";
export const ADataSurface6Subtle = "rgba(192, 178, 210, 1)";
export const ADataBorder1 = "rgba(51, 134, 224, 1)";
export const ADataBorder2 = "rgba(0, 91, 130, 1)";
export const ADataBorder3 = "rgba(199, 115, 0, 1)";
export const ADataBorder4 = "rgba(54, 141, 168, 1)";
export const ADataBorder5 = "rgba(51, 170, 95, 1)";
export const ADataBorder6 = "rgba(130, 105, 162, 1)";
export const ATransparent = "rgba(255, 255, 255, 0)";
export const AWhite = "rgba(255, 255, 255, 1)";
export const ANavRed = "rgba(195, 0, 0, 1)";
export const AGray50 = "rgba(242, 243, 245, 1)";
export const AGray100 = "rgba(236, 238, 240, 1)";
export const AGray200 = "rgba(224, 227, 230, 1)";
export const AGray300 = "rgba(203, 207, 213, 1)";
export const AGray400 = "rgba(170, 176, 186, 1)";
export const AGray500 = "rgba(131, 140, 154, 1)";
export const AGray600 = "rgba(103, 111, 123, 1)";
export const AGray700 = "rgba(82, 89, 98, 1)";
export const AGray800 = "rgba(66, 71, 79, 1)";
export const AGray900 = "rgba(35, 38, 42, 1)";
export const AGrayalpha50 = "rgba(38, 55, 89, 0.06)";
export const AGrayalpha100 = "rgba(18, 43, 68, 0.08)";
export const AGrayalpha200 = "rgba(17, 41, 64, 0.13)";
export const AGrayalpha300 = "rgba(7, 26, 54, 0.21)";
export const AGrayalpha400 = "rgba(5, 23, 51, 0.34)";
export const AGrayalpha500 = "rgba(2, 20, 49, 0.49)";
export const AGrayalpha600 = "rgba(2, 15, 34, 0.6)";
export const AGrayalpha700 = "rgba(1, 11, 24, 0.68)";
export const AGrayalpha800 = "rgba(3, 11, 22, 0.75)";
export const AGrayalpha900 = "rgba(2, 5, 9, 0.87)";
export const ABlue50 = "rgba(230, 240, 255, 1)";
export const ABlue100 = "rgba(204, 225, 255, 1)";
export const ABlue200 = "rgba(153, 195, 255, 1)";
export const ABlue300 = "rgba(102, 165, 244, 1)";
export const ABlue400 = "rgba(51, 134, 224, 1)";
export const ABlue500 = "rgba(0, 103, 197, 1)";
export const ABlue600 = "rgba(0, 86, 180, 1)";
export const ABlue700 = "rgba(0, 69, 156, 1)";
export const ABlue800 = "rgba(0, 52, 125, 1)";
export const ABlue900 = "rgba(0, 34, 82, 1)";
export const ARed50 = "rgba(255, 230, 230, 1)";
export const ARed100 = "rgba(255, 194, 194, 1)";
export const ARed200 = "rgba(246, 130, 130, 1)";
export const ARed300 = "rgba(242, 92, 92, 1)";
export const ARed400 = "rgba(222, 46, 46, 1)";
export const ARed500 = "rgba(195, 0, 0, 1)";
export const ARed600 = "rgba(173, 0, 0, 1)";
export const ARed700 = "rgba(140, 0, 0, 1)";
export const ARed800 = "rgba(92, 0, 0, 1)";
export const ARed900 = "rgba(38, 0, 0, 1)";
export const ADeepblue50 = "rgba(230, 241, 248, 1)";
export const ADeepblue100 = "rgba(204, 226, 240, 1)";
export const ADeepblue200 = "rgba(153, 196, 221, 1)";
export const ADeepblue300 = "rgba(102, 163, 196, 1)";
export const ADeepblue400 = "rgba(51, 128, 165, 1)";
export const ADeepblue500 = "rgba(0, 91, 130, 1)";
export const ADeepblue600 = "rgba(0, 80, 119, 1)";
export const ADeepblue700 = "rgba(0, 67, 103, 1)";
export const ADeepblue800 = "rgba(0, 52, 83, 1)";
export const ADeepblue900 = "rgba(0, 36, 58, 1)";
export const AGreen50 = "rgba(243, 252, 245, 1)";
export const AGreen100 = "rgba(204, 241, 214, 1)";
export const AGreen200 = "rgba(153, 222, 173, 1)";
export const AGreen300 = "rgba(102, 199, 134, 1)";
export const AGreen400 = "rgba(51, 170, 95, 1)";
export const AGreen500 = "rgba(6, 137, 58, 1)";
export const AGreen600 = "rgba(0, 124, 46, 1)";
export const AGreen700 = "rgba(0, 106, 35, 1)";
export const AGreen800 = "rgba(0, 85, 25, 1)";
export const AGreen900 = "rgba(0, 59, 15, 1)";
export const ALightblue50 = "rgba(235, 252, 255, 1)";
export const ALightblue100 = "rgba(216, 249, 255, 1)";
export const ALightblue200 = "rgba(181, 241, 255, 1)";
export const ALightblue300 = "rgba(151, 230, 255, 1)";
export const ALightblue400 = "rgba(124, 218, 248, 1)";
export const ALightblue500 = "rgba(102, 203, 236, 1)";
export const ALightblue600 = "rgba(76, 173, 205, 1)";
export const ALightblue700 = "rgba(54, 141, 168, 1)";
export const ALightblue800 = "rgba(35, 107, 125, 1)";
export const ALightblue900 = "rgba(19, 72, 82, 1)";
export const ALimegreen50 = "rgba(253, 255, 230, 1)";
export const ALimegreen100 = "rgba(249, 252, 204, 1)";
export const ALimegreen200 = "rgba(236, 243, 153, 1)";
export const ALimegreen300 = "rgba(217, 227, 102, 1)";
export const ALimegreen400 = "rgba(193, 203, 51, 1)";
export const ALimegreen500 = "rgba(162, 173, 0, 1)";
export const ALimegreen600 = "rgba(147, 158, 0, 1)";
export const ALimegreen700 = "rgba(127, 137, 0, 1)";
export const ALimegreen800 = "rgba(102, 110, 0, 1)";
export const ALimegreen900 = "rgba(71, 78, 0, 1)";
export const AOrange50 = "rgba(255, 249, 240, 1)";
export const AOrange100 = "rgba(255, 236, 204, 1)";
export const AOrange200 = "rgba(255, 215, 153, 1)";
export const AOrange300 = "rgba(255, 193, 102, 1)";
export const AOrange400 = "rgba(255, 170, 51, 1)";
export const AOrange500 = "rgba(255, 145, 0, 1)";
export const AOrange600 = "rgba(199, 115, 0, 1)";
export const AOrange700 = "rgba(168, 100, 0, 1)";
export const AOrange800 = "rgba(125, 76, 0, 1)";
export const AOrange900 = "rgba(82, 51, 0, 1)";
export const APurple50 = "rgba(239, 236, 244, 1)";
export const APurple100 = "rgba(224, 216, 233, 1)";
export const APurple200 = "rgba(192, 178, 210, 1)";
export const APurple300 = "rgba(161, 141, 187, 1)";
export const APurple400 = "rgba(130, 105, 162, 1)";
export const APurple500 = "rgba(99, 70, 137, 1)";
export const APurple600 = "rgba(82, 56, 116, 1)";
export const APurple700 = "rgba(65, 43, 93, 1)";
export const APurple800 = "rgba(48, 31, 70, 1)";
export const APurple900 = "rgba(31, 20, 47, 1)";
export const AShadowFocus = "0 0 0 3px rgba(0, 52, 125, 1)";
export const AShadowFocusInverted = "0 0 0 3px rgba(153, 195, 255, 1)";
export const AShadowXsmall = "0px 1px 3px 0px rgba(0, 0, 0, 0.15), 0px 0px 1px 0px rgba(0, 0, 0, 0.20)";
export const AShadowSmall = "0px 3px 8px 0px rgba(0, 0, 0, 0.10), 0px 1px 3px 0px rgba(0, 0, 0, 0.10), 0px 0px 1px 0px rgba(0, 0, 0, 0.18)";
export const AShadowMedium = "0px 5px 12px 0px rgba(0, 0, 0, 0.13), 0px 1px 3px 0px rgba(0, 0, 0, 0.10), 0px 0px 1px 0px rgba(0, 0, 0, 0.15)";
export const AShadowLarge = "0px 2px 5px 0px rgba(0, 0, 0, 0.15), 0px 10px 16px 0px rgba(0, 0, 0, 0.12), 0px 0px 1px 0px rgba(0, 0, 0, 0.12)";
export const AShadowXlarge = "0px 2px 5px 0px rgba(0, 0, 0, 0.15), 0px 10px 24px 0px rgba(0, 0, 0, 0.18), 0px 0px 1px 0px rgba(0, 0, 0, 0.08)";
export const ASpacing0 = "0";
export const ASpacing1 = "0.25rem";
export const ASpacing2 = "0.5rem";
export const ASpacing3 = "0.75rem";
export const ASpacing4 = "1rem";
export const ASpacing5 = "1.25rem";
export const ASpacing6 = "1.5rem";
export const ASpacing7 = "1.75rem";
export const ASpacing8 = "2rem";
export const ASpacing9 = "2.25rem";
export const ASpacing10 = "2.5rem";
export const ASpacing11 = "2.75rem";
export const ASpacing12 = "3rem";
export const ASpacing14 = "3.5rem";
export const ASpacing16 = "4rem";
export const ASpacing18 = "4.5rem";
export const ASpacing20 = "5rem";
export const ASpacing24 = "6rem";
export const ASpacing32 = "8rem";
export const ASpacing05 = "0.125rem";
export const ASpacing1Alt = "0.375rem";
