// @navikt/ds-react/cjs/typography/index.js
"use strict";
"use client";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Label = exports.Ingress = exports.Heading = exports.ErrorMessage = exports.Detail = exports.BodyShort = exports.BodyLong = void 0;
var BodyLong_1 = require("./BodyLong");
Object.defineProperty(exports, "BodyLong", { enumerable: true, get: function () { return __importDefault(BodyLong_1).default; } });
var BodyShort_1 = require("./BodyShort");
Object.defineProperty(exports, "BodyShort", { enumerable: true, get: function () { return __importDefault(BodyShort_1).default; } });
var Detail_1 = require("./Detail");
Object.defineProperty(exports, "Detail", { enumerable: true, get: function () { return __importDefault(Detail_1).default; } });
var ErrorMessage_1 = require("./ErrorMessage");
Object.defineProperty(exports, "ErrorMessage", { enumerable: true, get: function () { return __importDefault(ErrorMessage_1).default; } });
var Heading_1 = require("./Heading");
Object.defineProperty(exports, "Heading", { enumerable: true, get: function () { return __importDefault(Heading_1).default; } });
var Ingress_1 = require("./Ingress");
Object.defineProperty(exports, "Ingress", { enumerable: true, get: function () { return __importDefault(Ingress_1).default; } });
var Label_1 = require("./Label");
Object.defineProperty(exports, "Label", { enumerable: true, get: function () { return __importDefault(Label_1).default; } });
