// @navikt/ds-react/cjs/date/datepicker/DatePicker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatePicker = void 0;
const clsx_1 = __importDefault(require("clsx"));
const date_fns_1 = require("date-fns");
const react_1 = __importStar(require("react"));
const react_day_picker_1 = require("react-day-picker");
const util_1 = require("../../util");
const hooks_1 = require("../../util/hooks");
const useMergeRefs_1 = require("../../util/hooks/useMergeRefs");
const context_1 = require("../context");
const DateInput_1 = require("../parts/DateInput");
const DateWrapper_1 = require("../parts/DateWrapper");
const utils_1 = require("../utils");
const DatePickerStandalone_1 = __importDefault(require("./DatePickerStandalone"));
const Caption_1 = __importDefault(require("./parts/Caption"));
const DropdownCaption_1 = __importDefault(require("./parts/DropdownCaption"));
const HeadRow_1 = require("./parts/HeadRow");
const Row_1 = __importDefault(require("./parts/Row"));
const TableHead_1 = __importDefault(require("./parts/TableHead"));
const WeekNumber_1 = __importDefault(require("./parts/WeekNumber"));
/**
 * A component that allows users to select a date from a calendar.
 *
 * @see [ðŸ“ Documentation](https://aksel.nav.no/komponenter/core/datepicker)
 * @see ðŸ·ï¸ {@link DatePickerProps}
 *
 * @example
 * ```jsx
 *  const { inputProps, datepickerProps } = useMonthpicker({
 *    onMonthChange: console.log,
 *  });
 *
 *  return (
 *     <DatePicker {...datepickerProps} dropdownCaption>
 *       <DatePicker.Input
 *         {...inputProps}
 *         label="Velg dato"
 *       />
 *     </DatePicker>
 *  );
 * ```
 */
exports.DatePicker = (0, react_1.forwardRef)((_a, ref) => {
    var _b;
    var { children, locale = "nb", dropdownCaption, disabled = [], disableWeekends = false, showWeekNumber = false, selected, id, defaultSelected, className, wrapperClassName, open: _open, onClose, onOpenToggle, strategy, onWeekNumberClick } = _a, rest = __rest(_a, ["children", "locale", "dropdownCaption", "disabled", "disableWeekends", "showWeekNumber", "selected", "id", "defaultSelected", "className", "wrapperClassName", "open", "onClose", "onOpenToggle", "strategy", "onWeekNumberClick"]);
    const ariaId = (0, hooks_1.useId)(id);
    const [open, setOpen] = (0, react_1.useState)(_open !== null && _open !== void 0 ? _open : false);
    const wrapperRef = (0, react_1.useRef)(null);
    const mergedRef = (0, useMergeRefs_1.useMergeRefs)(wrapperRef, ref);
    const [selectedDates, setSelectedDates] = react_1.default.useState(defaultSelected);
    const mode = (_b = rest.mode) !== null && _b !== void 0 ? _b : "single";
    /**
     * @param newSelected Date | Date[] | DateRange | undefined
     */
    const handleSelect = (newSelected) => {
        var _a, _b, _c;
        setSelectedDates(newSelected);
        if (rest.mode === "single") {
            newSelected && ((_a = onClose === null || onClose === void 0 ? void 0 : onClose()) !== null && _a !== void 0 ? _a : setOpen(false));
        }
        else if (rest.mode === "range") {
            (newSelected === null || newSelected === void 0 ? void 0 : newSelected.from) && (newSelected === null || newSelected === void 0 ? void 0 : newSelected.to) && ((_b = onClose === null || onClose === void 0 ? void 0 : onClose()) !== null && _b !== void 0 ? _b : setOpen(false));
        }
        (_c = rest === null || rest === void 0 ? void 0 : rest.onSelect) === null || _c === void 0 ? void 0 : _c.call(rest, newSelected);
    };
    const DatePickerComponent = (react_1.default.createElement(react_day_picker_1.DayPicker, Object.assign({ locale: (0, utils_1.getLocaleFromString)(locale), mode: mode, onSelect: handleSelect, selected: selected !== null && selected !== void 0 ? selected : selectedDates, components: {
            Caption: dropdownCaption ? DropdownCaption_1.default : Caption_1.default,
            Head: TableHead_1.default,
            HeadRow: HeadRow_1.HeadRow,
            WeekNumber: WeekNumber_1.default,
            Row: Row_1.default,
        }, className: (0, clsx_1.default)("navds-date", className), classNames: {
            vhidden: "navds-sr-only",
        }, disabled: (day) => {
            return (disableWeekends && (0, date_fns_1.isWeekend)(day)) || (0, react_day_picker_1.isMatch)(day, disabled);
        }, weekStartsOn: 1, initialFocus: false, labels: utils_1.labels, modifiers: {
            weekend: (day) => disableWeekends && (0, date_fns_1.isWeekend)(day),
        }, modifiersClassNames: {
            weekend: "rdp-day__weekend",
        }, showWeekNumber: showWeekNumber, onWeekNumberClick: mode === "multiple" ? onWeekNumberClick : undefined, fixedWeeks: true, showOutsideDays: true }, (0, util_1.omit)(rest, ["onSelect"]))));
    return (react_1.default.createElement(context_1.DateContext.Provider, { value: {
            open: _open !== null && _open !== void 0 ? _open : open,
            onOpen: () => {
                setOpen((x) => !x);
                onOpenToggle === null || onOpenToggle === void 0 ? void 0 : onOpenToggle();
            },
            ariaId,
            defined: true,
        } },
        react_1.default.createElement("div", { ref: mergedRef, className: (0, clsx_1.default)("navds-date__wrapper", wrapperClassName) },
            children,
            react_1.default.createElement(DateWrapper_1.DateWrapper, { open: _open !== null && _open !== void 0 ? _open : open, anchor: wrapperRef.current, onClose: () => { var _a; return (_a = onClose === null || onClose === void 0 ? void 0 : onClose()) !== null && _a !== void 0 ? _a : setOpen(false); }, locale: locale, variant: mode, popoverProps: {
                    id: ariaId,
                    strategy,
                } }, DatePickerComponent))));
});
exports.DatePicker.Standalone = DatePickerStandalone_1.default;
exports.DatePicker.Input = DateInput_1.DatePickerInput;
exports.default = exports.DatePicker;
