// @navikt/ds-react/cjs/table/index.js
"use strict";
"use client";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Table = exports.TableRow = exports.TableHeaderCell = exports.TableHeader = exports.TableExpandableRow = exports.TableDataCell = exports.TableColumnHeader = exports.TableBody = void 0;
var Body_1 = require("./Body");
Object.defineProperty(exports, "TableBody", { enumerable: true, get: function () { return __importDefault(Body_1).default; } });
var ColumnHeader_1 = require("./ColumnHeader");
Object.defineProperty(exports, "TableColumnHeader", { enumerable: true, get: function () { return __importDefault(ColumnHeader_1).default; } });
var DataCell_1 = require("./DataCell");
Object.defineProperty(exports, "TableDataCell", { enumerable: true, get: function () { return __importDefault(DataCell_1).default; } });
var ExpandableRow_1 = require("./ExpandableRow");
Object.defineProperty(exports, "TableExpandableRow", { enumerable: true, get: function () { return __importDefault(ExpandableRow_1).default; } });
var Header_1 = require("./Header");
Object.defineProperty(exports, "TableHeader", { enumerable: true, get: function () { return __importDefault(Header_1).default; } });
var HeaderCell_1 = require("./HeaderCell");
Object.defineProperty(exports, "TableHeaderCell", { enumerable: true, get: function () { return __importDefault(HeaderCell_1).default; } });
var Row_1 = require("./Row");
Object.defineProperty(exports, "TableRow", { enumerable: true, get: function () { return __importDefault(Row_1).default; } });
var Table_1 = require("./Table");
Object.defineProperty(exports, "Table", { enumerable: true, get: function () { return __importDefault(Table_1).default; } });
